mantle/platform/qemu.go

```diff
@@ -1452,10 +1427,11 @@ func (builder *QemuBuilder) Exec() (*QemuInstance, error) {
 
 	}
 
-	// Set up QMP (currently used to switch boot order after first boot on aarch64)
-	qmpPath := filepath.Join(builder.tempdir, "qmp.sock")
+	// Set up QMP (currently used to switch boot order after first boot on aarch64.
+	// The qmp socket path must be unique to the instance.
+	inst.qmpSocketPath = filepath.Join(builder.tempdir, fmt.Sprintf("qmp-%d.sock", time.Now().UnixNano()))
```

Comment:
Just to make sure that my understanding is correct, `QemuBuilder` instances are per machine correct? Want to make sure there isn't a super rare race condition where we could still end up with two sockets of the same name.

Reply:
Correct, the instances are per machine.

Reply:
> Want to make sure there isn't a super rare race condition where we could still end up with two sockets of the same name/path.

After thinking about this, there is _one_ case where I can see a race condition happening:
- machine 1 defined
- someone copies the `QemuInstance` struct as machine 2
- machine 1 and machine 2 would share the same socket, disks, ignition....

In which case, the dev would have data corruption anyway.

Reply:
I'll see if I can find some time to dig a bit more through the code but my initial impression of the race in your example is the result of a gross misuse of the API and probably isn't worth solving for.

Reply:
Right, @arithx my example of the race would be an abuse of the API, but I was positing it for completeness in looking for `super race race conditions`.

Reply:
Just so I understand:

> Previously the qmp socket was hard-coded as a `qmp.sock` in the
instance's tmpdir. This is prone to race conditions (when one VM is torn
down and another is started using the same socket) or worse, cause flaky
and incorrect tests results for multiple VM/cluster scenerios.

Is this a *theoretical* concern or you saw a report from the Go race detector?

Each qemu instance has its own `builder.tempdir`.

I don't think anything should be copying the builder struct - if they were we should be sure to fix that.

That all said, the new code is better for sure!